metricbeat/module/kubernetes/event/event.go

```diff
@@ -0,0 +1,129 @@
+// The event MetricSet listens to events from Kubernetes API server and streams them to the output.
+// MetricSet implements the mb.PushMetricSet interface, and therefore does not rely on polling.
+type MetricSet struct {
+	mb.BaseMetricSet
+	watcher *Watcher
+}
+
+// New create a new instance of the MetricSet
+// Part of new is also setting up the configuration by processing additional
+// configuration entries if needed.
+func New(base mb.BaseMetricSet) (mb.MetricSet, error) {
+	logp.Experimental("The kubernetes event metricset is experimental")
+
+	config := defaultKuberentesEventsConfig()
+
+	err := base.Module().UnpackConfig(&config)
+	if err != nil {
+		return nil, fmt.Errorf("fail to unpack the kubernetes event configuration: %s", err)
+	}
+
+	var client *k8s.Client
+	if config.InCluster == true {
+		client, err = k8s.NewInClusterClient()
+		if err != nil {
+			return nil, fmt.Errorf("Unable to get in cluster configuration")
+		}
+	} else {
+		data, err := ioutil.ReadFile(config.KubeConfig)
+		if err != nil {
+			return nil, fmt.Errorf("read kubeconfig: %v", err)
+		}
+
+		// Unmarshal YAML into a Kubernetes config object.
+		var config k8s.Config
+		if err = yaml.Unmarshal(data, &config); err != nil {
+			return nil, fmt.Errorf("unmarshal kubeconfig: %v", err)
+		}
+		client, err = k8s.NewClient(&config)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	watcher := NewWatcher(client, config.SyncPeriod, config.Namespace)
+
+	return &MetricSet{
+		BaseMetricSet: base,
+		watcher:       watcher,
+	}, nil
+}
+
+// Run method provides the Kubernetes event watcher with a reporter with which events can be reported.
+func (m *MetricSet) Run(reporter mb.PushReporter) {
+	// Start event watcher
+	m.watcher.Run()
+
+	for {
+		select {
+		case <-reporter.Done():
+			m.watcher.Stop()
+			return
+		case msg := <-m.watcher.eventQueue:
+			// Ignore events that are deleted
+			if msg.Metadata.DeletionTimestamp == "" {
+				if msg.Metadata.DeletionTimestamp == "" {
+					reporter.Event(generateMapStrFromEvent(msg))
+				}
+			}
+		}
+	}
+}
+
+func generateMapStrFromEvent(eve *Event) common.MapStr {
+	return common.MapStr{
+		"timestamp": common.MapStr{
+			"first_occurrence": eve.FirstTimestamp.UTC(),
+			"last_occurrence":  eve.LastTimestamp.UTC(),
+		},
+		"message": eve.Message,
+		"reason":  eve.Reason,
+		"type":    eve.Type,
+		"count":   eve.Count,
+		"involved_object": common.MapStr{
+			"api_version":      eve.InvolvedObject.APIVersion,
+			"resource_version": eve.InvolvedObject.ResourceVersion,
+			"name":             eve.InvolvedObject.Name,
+			"kind":             eve.InvolvedObject.Kind,
+			"uid":              eve.InvolvedObject.UID,
+		},
+		"metadata": common.MapStr{
+			"timestamp": common.MapStr{
+				"created": eve.Metadata.CreationTimestamp,
+				"deleted": eve.Metadata.DeletionTimestamp,
+			},
+			"name":             eve.Metadata.Name,
+			"namespace":        eve.Metadata.Namespace,
```

Comment:
we would need to do this across all metricsets to establish consistency. lets do it in a follow-up PR all in one go.